source/fuzz/transformation_add_read_modify_write_atomic_instruction.cpp

```diff
@@ -0,0 +1,487 @@
+// Copyright (c) 2021 Mostafa Ashraf
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_add_read_modify_write_atomic_instruction.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationAddReadModifyWriteAtomicInstruction::
+    TransformationAddReadModifyWriteAtomicInstruction(
+        protobufs::TransformationAddReadModifyWriteAtomicInstruction message)
+    : message_(std::move(message)) {}
+
+TransformationAddReadModifyWriteAtomicInstruction::
+    TransformationAddReadModifyWriteAtomicInstruction(
+        uint32_t fresh_id, uint32_t pointer_id, uint32_t opcode,
+        uint32_t memory_scope_id, uint32_t memory_semantics_id_1,
+        uint32_t memory_semantics_id_2, uint32_t value_id,
+        uint32_t comparator_id,
+        const protobufs::InstructionDescriptor& instruction_to_insert_before) {
+  message_.set_fresh_id(fresh_id);
+  message_.set_pointer_id(pointer_id);
+  message_.set_opcode(opcode);
+  message_.set_memory_scope_id(memory_scope_id);
+  message_.set_memory_semantics_id_1(memory_semantics_id_1);
+  message_.set_memory_semantics_id_2(memory_semantics_id_2);
+  message_.set_value_id(value_id);
+  message_.set_comparator_id(comparator_id);
+
+  *message_.mutable_instruction_to_insert_before() =
+      instruction_to_insert_before;
+}
+
+bool TransformationAddReadModifyWriteAtomicInstruction::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // The result id must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // The pointer must exist and have a type.
+  auto pointer = ir_context->get_def_use_mgr()->GetDef(message_.pointer_id());
+  if (!pointer || !pointer->type_id()) {
+    return false;
+  }
+
+  // Read-Modify-Write atomic instructions are valid only.
+  if (pointer->opcode() == SpvOpAtomicLoad ||
+      pointer->opcode() == SpvOpAtomicStore) {
+    return false;
+  }
```

Comment:
I don't understand this.

Reply:
This transformation is responsible for adding atomic(read-modify-write) instructions only, so if the instruction is `SpvOpAtomicLoad` or `SpvOpAtomicStore` this is not acceptable, because we already supported them into this PRs [#4330](https://github.com/KhronosGroup/SPIRV-Tools/pull/4330) and [#4440 ](https://github.com/KhronosGroup/SPIRV-Tools/pull/4440).
What do you think @paulthomson?

Reply:
I don't think the opcode of the pointer matters, does it?

Reply:
Yes, you are right.
But in this case, we can't use this transformation for adding `SpvOpAtomicLoad` or `SpvOpAtomicStore`.

Reply:
Perhaps you meant to check `message_.opcode()`?

Reply:
Yes, you are right. Sorry for this mistake.

Reply:
Changed.

Reply:
However, I still don't think you need this. As with our other PRs, you can assert in the constructors that the opcode is one of the opcodes that we support. There is no need to have this in IsApplicable.

Reply:
Done.